src/main/java/com/google/collegeplanner/servlets/SchedulerServlet.java

```diff
@@ -0,0 +1,138 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.collegeplanner.servlets;
+
+import com.google.collegeplanner.data.Schedule;
+import com.google.collegeplanner.data.Section;
+import com.google.collegeplanner.data.SemesterScheduler;
+import com.google.gson.Gson;
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.ArrayList;
+import java.util.stream.Collectors;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import org.apache.http.client.utils.URIBuilder;
+import org.json.simple.JSONArray;
+import org.json.simple.JSONObject;
+import org.json.simple.parser.JSONParser;
+import org.json.simple.parser.ParseException;
+
+/** Servlet that returns a list of possible schedules containing the given courses.*/
+@WebServlet("/api/scheduler")
+public class SchedulerServlet extends BaseServlet {
+  private ArrayList<String> courseList;
+  private ArrayList<ArrayList<Section>> courses;
+
+  /**
+   * Organizes courses from POST request into a given number of semesters
+   */
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    JSONObject body;
+    JSONArray selectedClasses;
+    try {
+      body = getPostRequestBody(request);
+      selectedClasses = (JSONArray) body.get("selectedClasses");
+    } catch (NumberFormatException | ClassCastException | ParseException | NullPointerException e) {
+      respondWithError(
+          "Invalid body for POST request.", HttpServletResponse.SC_BAD_REQUEST, response);
+      return;
+    }
+
+    prepareLists(selectedClasses, response);
+
+    JSONObject schedules = getSchedules();
+    response.setContentType("application/json;");
+    response.getWriter().println(new Gson().toJson(schedules));
+  }
+
+  /**
+   * This method loads the courses and courseList ArrayLists with correct
+   * information. courses is loaded with the lists of sections for each
+   * course while courseList is loaded with the courseID's of the selected
+   * courses.
+   * @param classes The JSONArray of courseIds of the selected classes
+   * @param response The HttpServletResponse object.
+   */
+  private void prepareLists(JSONArray classes, HttpServletResponse response) throws IOException {
+    String courseId;
+    URI uri;
+    JSONArray jsonArray;
+    ApiUtil apiUtil = new ApiUtil();
+
+    courseList = new ArrayList<String>();
+    courses = new ArrayList<ArrayList<Section>>();
+    for (Object obj : classes) {
+      courseId = (String) obj;
+      courseList.add(courseId);
+      try {
+        uri = new URI("https://api.umd.io/v1/courses/" + courseId + "/sections");
```

Comment:
courseId should probably be escaped for safety

Reply:
what do you mean by escaped for safety?

Reply:
@savsa @ramyabuva there's one other java example for uri-escaping this right? can you point naaoli to it?

Reply:
I think the URI class automatically uri-escapes the input string, the only time we uri-escaped was in the Javascript

Reply:
Hmm, i'm not sure that's right. In this case, if `courseId` was `"foo/bar?baz="`, it could modify the url quite a bit, as it's just a raw string when you pass it in.

Reply:
https://github.com/googleinterns/step145-2020/blob/6485b74a3bfd38d8e4346a7b7f88bda00c253057/src/main/java/com/google/collegeplanner/servlets/DatastoreServlet.java#L141